File: c++/SingletonPattern/SafeLockSingleton.cpp
#include "stdafx.h"
#include "SafeLockSingleton.h"

mutex m_mutex1;
SafeLockSingleton *SafeLockSingleton::instance = NULL; //静态成员初始化

SafeLockSingleton::SafeLockSingleton()
{

}

SafeLockSingleton *SafeLockSingleton::GetInstance()
{
	m_mutex1.lock();
	if(instance==NULL)
	{
		instance = new SafeLockSingleton();
	}
	m_mutex1.unlock();
	return instance;
}

void SafeLockSingleton::Print()
{
	cout<<"线程安全的单例模式."<<endl;
}

File: csharp/SingletonPattern/SafeLockSingleton.cs

using System;
namespace SingletonPattern
{
    /// <summary>
    /// 线程安全的单例模式
    /// </summary>
    public sealed class SafeLockSingleton
    {
        private static SafeLockSingleton instance;
        private static readonly object syncFlag = new object();

        /// <summary>
        /// 私有的构造函数
        /// </summary>
        private SafeLockSingleton()
        {

        }

        public static SafeLockSingleton GetInstance()
        {
            lock (syncFlag)
            {
                if (instance == null)
                {
                    instance = new SafeLockSingleton();
                }
            }
            return instance;
        }

        public void Print()
        {
            Console.WriteLine("线程安全的单例模式.");
        }
    }
}
